### PR TITLE
Resolve login data load concurrency issues (allows ISOs to load at login)

### PR DESCRIPTION
**Pull request by sjd78**
Fixes: #1023, Fixes: #789 

Refactored the login sagas such that they load in an order that allows all required data to be present in the redux store prior to it being needed but other operations.  For example, a user's groups need to be loaded before permissions can be appropriately calculated.  Also, specifically for #1023, the set of storage domains for a user needs to be loaded before the ISOs can be properly fetched.

Details:
  - Login/logout sagas better encapsulated in sagas/login.js

  - Moved server config/options to sagas/server-configs.js

  - Data load concurrency issues have been resolved.  Fetches happen in the order needed to ensure all required data is loaded prior to it being needed.

  - Each stage is dependent on (gated by) the previous stage's values:
    - Login information
    - User information
    - Server stored configurations
    - Data with no other pre-reqs
    - Data requiring user groups (and permissions/permits)
    - Data requiring storage domains

  - When possible at each stage, fetches are run as parallel tasks.

  - Added "appConfigured" flag to config reducer.  The login process will set the `config` reducer's `appConfigured` flag once enough configuration data has been loaded and the App can properly render.

  - The user domain is now part of the login action and doesn't require a separate action.

Updated configuration of redux dev tools:
  - Configure REDUX_DEVTOOLS to keep 100 actions in history.  The default is 50 but that doesn't let us see the first few actions at App start-up time.  100 works better for now.

Updated trace logging to make it easier to trace multiple requests happening in parallel:
  - Trace logging on the transport has been enhanced to make it easier to match up requests and responses looking at console output
  - `callExternalAction` debug logging formatting updated to match that of the transport logging

Additional refactoring work:
  - Reorganize sagas/utils imports
  - Move `doCheckTokenExpired` from sagas/login to sagas/utils
  - Sagas that need to query multiple times now use `delayInMsSteps()`
    to generate query intervals
  - Remove unused 'shallowFetch' from getAllTemplates() action

### CODE REVIEW

**Comment by isaranova** [comment]
LGTM

File: src/ovirtapi/transport.js [hunk 1]
```diff
@@ -57,27 +57,32 @@ type GetRequestType = { url: string, custHeaders?: Object}
 type InputRequestType = { url: string, input: string, contentType?: string }
 type DeleteRequestType = { url: string, custHeaders?: Object }
 
+let getCounter = 0
+const logHeaders = (headers) => JSON.stringify({ ...headers, 'Authorization': '*****' })
+
 function httpGet ({ url, custHeaders = {} }: GetRequestType): Promise<Object> {
-  logger.log(`_httpGet start: url="${url}"`)
+  const myCounter = getCounter++
```
Comment (bond95) [inline src/ovirtapi/transport.js]: I don't see any usage of this, so think it can be removed.

File: src/sagas/login.js [hunk 2]
```diff
@@ -37,26 +29,37 @@ import {
...
+import { fetchDataCentersAndStorageDomains, fetchIsoFiles } from './storageDomains'
 
 /**
  * Perform login checks, and if they pass, perform initial data loading
  */
-export function* login (action) {
-  const { payload: { token, userId, credentials: { username } } } = action
+function* login (action) {
+  const { payload: { token, userId, username, domain } } = action
+  console.group('Login Verification')
```
Comment (bond95) [inline src/sagas/login.js]: Nice, I like using of groups.

File: src/sagas/login.js [hunk 3]
```diff
@@ -173,106 +138,83 @@ function* checkOvirtApiVersion (oVirtMeta) {
   return passed
 }
 
-export function* logout () {
-  if (AppConfiguration.applicationLogoutURL && AppConfiguration.applicationLogoutURL.length > 0) {
-    window.location.href = AppConfiguration.applicationLogoutURL
+function composeIncompatibleOVirtApiVersionMessage (oVirtMeta) {
```
Comment (bond95) [inline src/sagas/login.js]: I think this function must be in `utils` too

File: src/ovirtapi/transport.js (hunk 1 above)
Comment (sjd78) [reply src/ovirtapi/transport.js]: It's being used by the logger function lower down.  I added it so the request and responses can be matched in the log when fetches are being run in parallel.

File: src/sagas/login.js (hunk 3 above)
Comment (sjd78) [reply src/sagas/login.js]: Perhaps, but it is only ever used by the `login()`, and only if the version check fails.  I don't see the function being used anywhere else, so I went with keeping it here.

File: src/ovirtapi/transport.js (hunk 1 above)
Comment (bond95) [reply src/ovirtapi/transport.js]: Oh, yes, I missed it, thank you.

**Comment by michalskrivanek** [comment]
seems to me there are no fundamental comment. Anything to wait for, @bond95 ?

**Comment by michalskrivanek** [comment]
@sjd78 ready to be merged?

File: src/actions/index.js [hunk 4]
```diff
@@ -39,6 +40,13 @@ export * from './console'
 export * from './userMessages'
 export * from './disks'
 
+export function appConfigured () {
+  return {
+    type: APP_CONFIGURED,
+    payload: {},
```
Comment (bond95) [inline src/actions/index.js]: From time to time if payload is empty we just don't write it (like in https://github.com/oVirt/ovirt-web-ui/blob/master/src/actions/storageDomains.js). I like that style more than writing unnecessary payload.

File: src/actions/templates.js [hunk 5]
```diff
@@ -13,11 +13,9 @@ export function setTemplates (templates) {
...
 }
 
-export function getAllTemplates ({ shallowFetch = false }) {
+export function getAllTemplates () {
   return {
     type: GET_ALL_TEMPLATES,
-    payload: {
-      shallowFetch,
-    },
+    payload: {},
```
Comment (bond95) [inline src/actions/templates.js]: Same here.

File: src/ovirtapi/transport.js [hunk 6]
```diff
@@ -57,27 +57,32 @@ type GetRequestType = { url: string, custHeaders?: Object}
...
     'Authorization': `Bearer ${_getLoginToken()}`,
     'Accept-Language': AppConfiguration.queryParams.locale, // can be: undefined, empty or string
     'Filter': Selectors.getFilter(),
-    'Accept': 'application/json',
-  }, custHeaders)
-  logger.log(`_httpGet: url="${url}", headers="${JSON.stringify(headers)}"`)
+    ...custHeaders,
+  }
 
+  logger.log(`http GET[${myCounter}] -> url: "${url}", headers: ${logHeaders(headers)}`)
```
Comment (bond95) [inline src/ovirtapi/transport.js]: As I see in other files `logger` was rewritten to `console` is there any reason to leave it here?

File: src/ovirtapi/transport.js [hunk 7]
```diff
@@ -57,27 +57,32 @@ type GetRequestType = { url: string, custHeaders?: Object}
...
   return $.ajax(url, {
     type: 'GET',
     headers,
   })
     .then((data: Object): Object => {
       notifyStop(requestId)
+      logger.log(`http GET[${myCounter}] <- data:`, data)
```
Comment (bond95) [inline src/ovirtapi/transport.js]: Same

File: src/sagas/login.js (hunk 3 above)
Comment (bond95) [reply src/sagas/login.js]: As well as `doCheckTokenExpired`, that used only in `index.js`. So I still think that better move it to `utils`.

File: src/actions/index.js (hunk 4 above)
Comment (sjd78) [reply src/actions/index.js]: `storageDomains.js` contains the only 2 actions in the app that don't have a payload.  And for the sagas attached to those actions, they push a payload into the action, or define their own, before calling the external action.  I'll leave the `payload: {}` here.

Comment (bond95) [reply src/actions/index.js]: Yes, but still this unnecessary string, so why not to start refactor code from here, and remove it completely in the future? Even in sagas examples they don't use `payload: {}` where it not needed. https://github.com/redux-saga/redux-saga/blob/master/examples/shopping-cart/src/actions/index.js

Comment (sjd78) [reply src/actions/index.js]: Ok, then file a separate issue.  I'm not going to make any further changes to the actions in this PR.

File: src/sagas/login.js (hunk 3 above)
Comment (sjd78) [reply src/sagas/login.js]: Any `doCheckTokenExpired` and `callExternalAction` work should probably be done under the context of #991.  I'll leave tweaking these functions for that PR.

**Pull request merged**